syntax = "proto3";

package product;

service ProductService {
  rpc GetProduct(ProductRequest) returns (ProductResponse) {}
  rpc GetProductInventory(ProductRequest) returns (ProductInventoryResponse) {}
  rpc CheckProductAvailability(ProductAvailabilityRequest) returns (ProductAvailabilityResponse) {}
  rpc ReserveInventory(InventoryRequest) returns (InventoryResponse);
  rpc ReleaseInventory(InventoryRequest) returns (InventoryResponse); 
  rpc ConfirmInventory(InventoryRequest) returns (InventoryResponse);
  rpc CheckAndReserveInventory(InventoryRequest) returns (InventoryResponse);
}

message ProductRequest {
  string product_id = 1;
}

message ProductResponse {
  string product_id = 1;
  string title = 2;
  double price = 3;
}

message ProductInventoryResponse {
  string product_id = 1;
  int32 stock = 2;
  int32 stock_reserved = 3;
  int32 available_stock = 4;
}

message ProductAvailabilityRequest {
  string product_id = 1;
  int32 quantity = 2;
}

message ProductAvailabilityResponse {
  bool available = 1;
  string message = 2;
} 

message InventoryRequest {
  string product_id = 1;
  int32 quantity = 2;
}

message InventoryResponse {
  bool success = 1;
  string message = 2;
}